libs/x-content/src/test/java/org/elasticsearch/common/xcontent/InstantiatingObjectParserTests.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+
+public class InstantiatingObjectParserTests extends ESTestCase {
+
+    public static class NoAnnotations {
+        final int a;
+        final String b;
+        final long c;
+
+        public NoAnnotations() {
+            this(1, "2", 3);
+        }
+
+        private NoAnnotations(int a) {
+            this(a, "2", 3);
+        }
+
+        public NoAnnotations(int a, String b) {
+            this(a, b, 3);
+        }
+
+        public NoAnnotations(int a, long c) {
+            this(a, "2", c);
+        }
+
+        public NoAnnotations(int a, String b, long c) {
+            this.a = a;
+            this.b = b;
+            this.c = c;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            NoAnnotations that = (NoAnnotations) o;
+            return a == that.a &&
+                c == that.c &&
+                Objects.equals(b, that.b);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(a, b, c);
+        }
+    }
+
+    public void testNoAnnotation() throws IOException {
+        InstantiatingObjectParser<NoAnnotations, Void> parser = new InstantiatingObjectParser<>("foo", NoAnnotations.class);
+        parser.declareInt(constructorArg(), new ParseField("a"));
+        parser.declareString(constructorArg(), new ParseField("b"));
+        parser.declareLong(constructorArg(), new ParseField("c"));
+        parser.finalizeFields();
```

Comment:
It'd be kind of neat to have a test that uses the two-arg ctor or the one arg ctor. And maybe one that sets a member after calling the ctor just for paranoia's sake.